README.md
# flask_template
Cookiecutter Template for a Flask Webapp


CHANGELOG.rst
.. _flask-template-changelog:

==========
Change Log
==========

This document records the main changes to the flask-template code.


.. _changelog-0.1.0:

0.1.0 (2018-08-16)
------------------

Added
^^^^^
* Initial release.
* Includes Travis CI, Read The Docs, Coverage, and Bumpversion integrations.
* Includes a logger and configuration library.
* Examples for Sphinx documentation and pytest.



{{cookiecutter.repo_name}}/python/{{cookiecutter.package_name}}/tests/conftest.py
# !usr/bin/env python
# -*- coding: utf-8 -*-
#
# Licensed under a 3-clause BSD license.
#
# @Author: Brian Cherinka
# @Date:   2018-08-16 11:43:42
# @Last modified by:   Brian Cherinka
# @Last Modified time: 2018-08-16 23:09:00

from __future__ import print_function, division, absolute_import
import pytest
from flask import url_for
from {{cookiecutter.package_name}}.app import create_app


@pytest.fixture
def app():
    app = create_app()
    return app


def test_app(client):
    assert client.get(url_for('index.home')).status_code == 200


def test_404error(client):
    assert client.get('help').status_code == 404


def test_ok_status(client):
    res = client.get(url_for('index.status'))
    assert res.json == {'status': 'ok'}



{{cookiecutter.repo_name}}/python/{{cookiecutter.package_name}}/settings.py
# !usr/bin/env python
# -*- coding: utf-8 -*-
#
# Licensed under a 3-clause BSD license.
#
# @Author: Brian Cherinka
# @Date:   2018-08-16 11:14:09
# @Last modified by:   Brian Cherinka
# @Last Modified time: 2018-08-16 16:05:37

from __future__ import print_function, division, absolute_import
import os

NAME = '{{cookiecutter.package_name}}'.upper()


class Config(object):
    SECRET_KEY = os.environ.get(f'{NAME}_SECRET', os.urandom(24))
    FLASK_SECRET = SECRET_KEY
    APP_DIR = os.path.abspath(os.path.dirname(__file__))  # This directory
    APP_BASE = os.environ.get(f'{NAME}_BASE', '{{cookiecutter.package_name}}')
    projroot = os.path.abspath(os.path.join(APP_DIR, os.pardir, os.pardir))
    PROJECT_ROOT = os.environ.get(f'{NAME}_DIR', projroot)
    os.environ['FLASK_APP'] = os.path.join(APP_DIR, 'app.py')


class ProdConfig(Config):
    """Production configuration."""
    ENV = 'production'
    DEBUG = False
    os.environ['FLASK_ENV'] = ENV


class DevConfig(Config):
    """Development configuration."""
    ENV = 'development'
    DEBUG = True
    os.environ['FLASK_ENV'] = ENV


class TestConfig(Config):
    TESTING = True
    DEBUG = True


class CustomConfig(object):
    ''' Project specific configuration.  Always gets appended to an above Config class '''
    pass


{{cookiecutter.repo_name}}/python/{{cookiecutter.package_name}}/error_handlers.py
# !usr/bin/env python
# -*- coding: utf-8 -*-
#
# Licensed under a 3-clause BSD license.
#
# @Author: Brian Cherinka
# @Date:   2018-08-16 22:46:56
# @Last modified by:   Brian Cherinka
# @Last Modified time: 2018-08-16 22:50:48

from __future__ import print_function, division, absolute_import
from flask import request, current_app as app
from flask import Blueprint, render_template


errors = Blueprint('error_handlers', __name__)


def make_error_page(app, name, code, data=None, exception=None):
    ''' creates the error page dictionary for web errors '''
    shortname = name.lower().replace(' ', '_')
    error = {}
    error['title'] = 'Marvin | {0}'.format(name)
    error['page'] = request.url
    error['data'] = data
    error['name'] = name
    error['code'] = code
    error['message'] = exception.description if exception and hasattr(exception, 'description') else None
    app.logger.error('{0} Exception {1}'.format(name, error))
    return render_template('errors/{0}.html'.format(shortname), **error), code


@errors.app_errorhandler(404)
def page_not_found(error):
    ''' custom 404 page not found error '''
    name = 'Page Not Found'
    return make_error_page(app, name, 404, exception=error)


{{cookiecutter.repo_name}}/python/{{cookiecutter.package_name}}/app.py
# !usr/bin/env python
# -*- coding: utf-8 -*-
#
# Licensed under a 3-clause BSD license.
#
# @Author: Brian Cherinka
# @Date:   2018-05-15 10:51:48
# @Last modified by:   Brian Cherinka
# @Last Modified time: 2018-08-16 23:00:23

from __future__ import print_function, division, absolute_import
import os
from flask import Flask
from {{cookiecutter.package_name}}.controllers.index import index
from {{cookiecutter.package_name}}.settings import ProdConfig, DevConfig, CustomConfig
from {{cookiecutter.package_name}}.error_handlers import errors


# set the package name and version
__name__ = '{{cookiecutter.package_name}}'
__version__ = '{{cookiecutter.version}}'


def create_app(debug=None, local=None, object_config=None):
    ''' main web application factory '''

    # set a base name
    base = os.environ.get(f'{__name__.upper()}_BASE', __name__)

    # initiliaze the application
    app = Flask(__name__, static_url_path=f'/{base}/static')
    # set a url prefix
    url_prefix = f'/{__name__}' if local else f'/{base}'

    # ----------------------------------
    # Load the appropriate Flask configuration object for debug or production
    if not object_config:
        if app.debug or local:
            app.logger.info('Loading Development Config!')
            object_config = type('Config', (DevConfig, CustomConfig), dict())
        else:
            app.logger.info('Loading Production Config!')
            object_config = type('Config', (ProdConfig, CustomConfig), dict())
    app.config.from_object(object_config)

    # ----------------------------------
    # Registration
    register_api(app)
    register_extensions(app, app_base=base)
    register_blueprints(app, url_prefix=url_prefix)

    return app


def register_blueprints(app, url_prefix=None):
    ''' Register the Flask Blueprints used '''

    app.register_blueprint(index, url_prefix=url_prefix)
    app.register_blueprint(errors)


def register_extensions(app, app_base=None):
    ''' Register the Flask Extensions used '''


def register_api(app, api=None):
    ''' Register the Flask API routes used '''
    pass



